ACM#9987: HCP: Document Agent provisioning flow for non BM machines (#6073)
open-cluster-management/rhacm-docs: clusters/hosted_control_planes/create_cluster_non_bm.adoc

** On the *Node pools* page, the namespace contains the hosts for the node pool. If you created a host inventory by using the console, the console creates a dedicated namespace.

** On the *Networking* page, you select an API server publishing strategy. The API server for the hosted cluster can be exposed either by using an existing load balancer or as a service of the `NodePort` type. A DNS entry must exist for the `api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN}` setting that points to the destination where the API server can be reached. This entry can be a record that points to one of the nodes in the management cluster or a record that points to a load balancer that redirects incoming traffic to the Ingress pods.

. Review your entries and click **Create**.

+
The **Hosted cluster** view is displayed.

. Monitor the deployment of the hosted cluster in the **Hosted cluster** view. If you do not see information about the hosted cluster, ensure that **All Clusters** is selected, and click the cluster name.

> . Wait until the control plane components are ready. This process can take a few minutes.

Reviewer: Wait is not really a step, so that should be with the other steps. I will play with it.

[#creating-a-hosted-cluster-non-bm]
= Creating a hosted cluster on non bare metal agent machines

You can create a hosted cluster or import one. For instructions to import a hosted cluster, see _Importing a hosted cluster_.

. Verify that you have a default storage class configured for your cluster. Otherwise, you might end up with pending PVCs. Enter the following commands, replacing any example variables with your information:

+
----
export CLUSTERS_NAMESPACE="clusters"
export HOSTED_CLUSTER_NAME="example"
export HOSTED_CONTROL_PLANE_NAMESPACE="${CLUSTERS_NAMESPACE}-${HOSTED_CLUSTER_NAME}" <1>
export BASEDOMAIN="krnl.es"
export PULL_SECRET_FILE=$PWD/pull-secret
export MACHINE_CIDR=192.168.122.0/24
export ETCD_STORAGE="lvm-storageclass"
export OCP_RELEASE=4.14.0-multi
export SSH_KEY=${HOME}/.ssh/id_rsa.pub
oc create ns ${HOSTED_CONTROL_PLANE_NAMESPACE}

hcp create cluster agent \
    --name=${HOSTED_CLUSTER_NAME} \
    --pull-secret=${PULL_SECRET_FILE} \

>     --agent-namespace=${HOSTED_CONTROL_PLANE_NAMESPACE} \

Reviewer 2: If a note could be added here to ensure this is the namespace where your `Agent` CRs exist (which can be found by `oc get agent -n <this namespace>`

Reviewer 2: Might also be nice to have a link or reference to how they can create Agents too

Author: @CrystalChun Do you have any resource link handy to provide it here? I couldn't find one

Reviewer 2: Yes here you go! https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.9/html/clusters/cluster_mce_overview#add-host-host-inventory 
Will defer to you and your team for where to best link this!

[#creating-a-hosted-cluster-non-bm]
= Creating a hosted cluster on non bare metal agent machines

You can create a hosted cluster or import one. For instructions to import a hosted cluster, see _Importing a hosted cluster_.

. Verify that you have a default storage class configured for your cluster. Otherwise, you might end up with pending PVCs. Enter the following commands, replacing any example variables with your information:

+
----
export CLUSTERS_NAMESPACE="clusters"
export HOSTED_CLUSTER_NAME="example"
export HOSTED_CONTROL_PLANE_NAMESPACE="${CLUSTERS_NAMESPACE}-${HOSTED_CLUSTER_NAME}" <1>
export BASEDOMAIN="krnl.es"
export PULL_SECRET_FILE=$PWD/pull-secret
export MACHINE_CIDR=192.168.122.0/24
export ETCD_STORAGE="lvm-storageclass"

> export OCP_RELEASE=4.14.0-multi

Reviewer 3: I would change this by an official stable 4.14 version (e.g `4.14.13-x86_64`)

Reviewer: How we are writing these, to avoid creating bugs when we can, is with a variable:

`OCP_RELEASE=4.x.0-multi`

With either a call out or a simple line that says, "Replace 4.x.0 with the supported version" (or something like that, there are examples in the doc that we recently changed).

export SSH_KEY=${HOME}/.ssh/id_rsa.pub
oc create ns ${HOSTED_CONTROL_PLANE_NAMESPACE}

hcp create cluster agent \
    --name=${HOSTED_CLUSTER_NAME} \
    --pull-secret=${PULL_SECRET_FILE} \
    --agent-namespace=${HOSTED_CONTROL_PLANE_NAMESPACE} \ <2>
    --base-domain=${BASEDOMAIN} \
    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN} \
    --etcd-storage-class=${ETCD_STORAGE} \
    --ssh-key  ${SSH_KEY} \
    --namespace ${CLUSTERS_NAMESPACE} \
    --control-plane-availability-policy SingleReplica \
    --release-image=quay.io/openshift-release-dev/ocp-release:${OCP_RELEASE} \
    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN}
----

+

> <1> Typically, the namespace is created by the HyperShift Operator, but agent cluster creation generates a cluster API provider role that needs the namespace to already exist.
> <2> Ensure that agents are available in the `${HOSTED_CONTROL_PLANE_NAMESPACE}`. You can find the list of agents by using the `oc get agent -n ${HOSTED_CONTROL_PLANE_NAMESPACE}` command.

Reviewer 2: Sorry, I realize now that having a note in the command the user would run is not such a great idea 😅 Suggestion to merge it into <1> but feel free to do what you think is best!
```suggestion
<1> Typically, the namespace is created by the HyperShift Operator, but agent cluster creation generates a cluster API provider role that needs the namespace to already exist. In the command below, it is also used as the `agent-namespace` flag. Ensure that agents are available in the `${HOSTED_CONTROL_PLANE_NAMESPACE}`. You can find the list of agents by using the `oc get agent -n ${HOSTED_CONTROL_PLANE_NAMESPACE}` command.
```

Reviewer 2: Checking if this change was meant to occur, if it wasn't, feel free to ignore this comment!

[#creating-a-hosted-cluster-non-bm]
= Creating a hosted cluster on non bare metal agent machines

You can create a hosted cluster or import one. For instructions to import a hosted cluster, see _Importing a hosted cluster_.

. Verify that you have a default storage class configured for your cluster. Otherwise, you might end up with pending PVCs. Enter the following commands, replacing any example variables with your information:

+
----

> export CLUSTERS_NAMESPACE="clusters"
> export HOSTED_CLUSTER_NAME="example"
> export HOSTED_CONTROL_PLANE_NAMESPACE="${CLUSTERS_NAMESPACE}-${HOSTED_CLUSTER_NAME}"
> export BASEDOMAIN="krnl.es"
> export PULL_SECRET_FILE=$PWD/pull-secret
> export MACHINE_CIDR=192.168.122.0/24
> export ETCD_STORAGE="lvm-storageclass"
> export OCP_RELEASE=4.14.13-x86_64
> export SSH_KEY=${HOME}/.ssh/id_rsa.pub
> oc create ns ${HOSTED_CONTROL_PLANE_NAMESPACE}
> 
> hcp create cluster agent \
>     --name=${HOSTED_CLUSTER_NAME} \
>     --pull-secret=${PULL_SECRET_FILE} \
>     --agent-namespace=${HOSTED_CONTROL_PLANE_NAMESPACE} \ <1>
>     --base-domain=${BASEDOMAIN} \
>     --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN} \
>     --etcd-storage-class=${ETCD_STORAGE} \
>     --ssh-key  ${SSH_KEY} \
>     --namespace ${CLUSTERS_NAMESPACE} \
>     --control-plane-availability-policy SingleReplica \
>     --release-image=quay.io/openshift-release-dev/ocp-release:${OCP_RELEASE} \
>     --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN}
> ----
> 
> +
> <1> Required by the agent cluster API provider and must already exist. You can use the `--agent-namespace` flag in the `hcp create cluster` command. Ensure that agents are available in this namespace. You can find the list of agents by using the `oc get agent -n ${HOSTED_CONTROL_PLANE_NAMESPACE}` command.

Reviewer 2: ```suggestion
export CLUSTERS_NAMESPACE="clusters"
export HOSTED_CLUSTER_NAME="example"
export HOSTED_CONTROL_PLANE_NAMESPACE="${CLUSTERS_NAMESPACE}-${HOSTED_CLUSTER_NAME}" <1>
export BASEDOMAIN="krnl.es"
export PULL_SECRET_FILE=$PWD/pull-secret
export MACHINE_CIDR=192.168.122.0/24
export ETCD_STORAGE="lvm-storageclass"
export OCP_RELEASE=4.14.13-x86_64
export SSH_KEY=${HOME}/.ssh/id_rsa.pub
oc create ns ${HOSTED_CONTROL_PLANE_NAMESPACE}

hcp create cluster agent \
    --name=${HOSTED_CLUSTER_NAME} \
    --pull-secret=${PULL_SECRET_FILE} \
    --agent-namespace=${HOSTED_CONTROL_PLANE_NAMESPACE} \
    --base-domain=${BASEDOMAIN} \
    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN} \
    --etcd-storage-class=${ETCD_STORAGE} \
    --ssh-key  ${SSH_KEY} \
    --namespace ${CLUSTERS_NAMESPACE} \
    --control-plane-availability-policy SingleReplica \
    --release-image=quay.io/openshift-release-dev/ocp-release:${OCP_RELEASE} \
    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN}
----

+
<1> Required by the agent cluster API provider and must already exist. Use this as the `--agent-namespace` flag in the `hcp create cluster` command. Ensure that agents are available in this namespace. You can find the list of agents by using the `oc get agent -n ${HOSTED_CONTROL_PLANE_NAMESPACE}` command.
```

konnectivity-agent-c5485c9df-jsm9s               1/1     Running   0          4m13s
konnectivity-server-85dc754888-7z8vm             1/1     Running   0          4m13s
kube-apiserver-db5fb5549-zlvpq                   3/3     Running   0          4m13s
kube-controller-manager-5fbf7b7b7b-mrtjj         1/1     Running   0          90s
kube-scheduler-776c59d757-kfhv6                  1/1     Running   0          3m12s
machine-approver-c6b947895-lkdbk                 1/1     Running   0          4m13s
oauth-openshift-787b87cff6-trvd6                 2/2     Running   0          87s
olm-operator-69c4657864-hxwzk                    2/2     Running   0          2m50s
openshift-apiserver-67f9d9c5c7-c9bmv             2/2     Running   0          89s
openshift-controller-manager-5899fc8778-q89xh    1/1     Running   0          2m51s
openshift-oauth-apiserver-569c78c4d-568v8        1/1     Running   0          2m52s
packageserver-ddfffb8d7-wlz6l                    2/2     Running   0          2m50s
redhat-marketplace-catalog-7dd77d896-jtxkd       1/1     Running   0          2m51s
redhat-operators-catalog-d66b5c965-qwhn7         1/1     Running   0          2m51s

> ----

Reviewer 4: This is definitely an output that is too long and can be replaced by a sentence telling the user to check the `STATUS` of their pods, if they need verification.

[#hosted-create-non-bare-metal-console]

> == Creating a hosted cluster on non bare metal agent machines by using the console

Reviewer 4: We have some topics that document console processes, but we generally try to avoid documenting the console. I'm not sure if this would be considered too much, what do you think @swopebe?

Reviewer: Yes, we don't write UI steps unless 1. We have to document how to use a third-party console (Grafana) or 2. The console is not intuitive.

I think by using "follow the prompts" that this may be as concise as it can be, not sure the user could go to "create cluster" without these bold steps. 

I see some changes that need to be made, so I will review this in the afternoon, as well.

Author: IMO, since we cover creating hosted clusters using console for bare metal, the console procedure would make sense here for non bare metal as well.

Author: > Yes, we don't write UI steps unless 1. We have to document how to use a third-party console (Grafana) or 2. The console is not intuitive.
> 
> I think by using "follow the prompts" that this may be as concise as it can be, not sure the user could go to "create cluster" without these bold steps.
> 
> I see some changes that need to be made, so I will review this in the afternoon, as well.

Sure thanks

. Open the {ocp-short} web console and log in by entering your administrator credentials. For instructions to open the console, see link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html/web_console/web-console[Accessing the web console] in the {ocp-short} documentation.

. In the console header, ensure that **All Clusters** is selected.

> . Click **Infrastructure -> Clusters**.

Reviewer 4: ```suggestion
. Click **Infrastructure > Clusters**.
```

We just use a simple `>` for these, make sure to replace them in the rest of the PR as well.
